nixops/resources/vpc.py

```diff
@@ -0,0 +1,275 @@
+        nixops.resources.ResourceState.__init__(self, depl, name, id)
+        self._client = None
+        self._state = StateDict(depl, id)
+        self.vpc_id = self._state.get('vpcId', None)
+        self.handle_create = Handler(['cidrBlock', 'region', 'instanceTenancy'], handle=self.realize_create_vpc)
+        self.handle_dns = Handler(['enableDnsHostnames', 'enableDnsSupport'], after=[self.handle_create]
+                                  , handle=self.realize_dns_config)
+        self.handle_classic_link = Handler(['enableClassicLink'], after=[self.handle_create]
+                                           , handle=self.realize_classic_link_change)
+        self.handle_associate_ipv6_cidr_block = Handler(
+            ['amazonProvidedIpv6CidrBlock'],
+            after=[self.handle_create],
+            handle=self.realize_associate_ipv6_cidr_block)
+        self.handle_tag_update = Handler(['tags'], after=[self.handle_create], handle=self.realize_update_tag)
+
+    def show_type(self):
+        s = super(VPCState, self).show_type()
+        region = self._state.get('region', None)
+        if region: s = "{0} [{1}]".format(s, region)
+        return s
+
+    @property
+    def resource_id(self):
+        return self._state.get('vpcId', None)
+
+    def prefix_definition(self, attr):
+        return {('resources', 'vpc'): attr}
+
+    def get_physical_spec(self):
+        return { 'vpcId': self._state.get('vpcId', None)}
+
+    def get_definition_prefix(self):
+        return "resources.vpc."
+
+    def connect(self):
+        if self._client: return
+        assert self._state['region']
+        (access_key_id, secret_access_key) = nixops.ec2_utils.fetch_aws_secret_key(self.access_key_id)
+        self._client = boto3.session.Session().client('ec2', region_name=self._state['region'], aws_access_key_id=access_key_id, aws_secret_access_key=secret_access_key)
+
+    def _destroy(self):
+        if self.state != self.UP: return
+        self.connect()
+        self.log("destroying vpc {0}...".format(self._state['vpcId']))
+        try:
+            self._client.delete_vpc(VpcId=self._state['vpcId'])
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == 'InvalidVpcID.NotFound':
+                self.warn("vpc {0} was already deleted".format(self._state['vpcId']))
+            else:
+                raise e
+
+        with self.depl._db:
+            self.state = self.MISSING
+            self._state['vpcId'] = None
+            self._state['region'] = None
+            self._state['cidrBlock'] = None
+            self._state['instanceTenancy'] = None
+            self._state['enableDnsSupport'] = None
+            self._state['enableDnsHostname'] = None
+            self._state['enableVpcClassicLink'] = None
+
+    def create_after(self, resources, defn):
+        return {r for r in resources if
+                isinstance(r, nixops.resources.elastic_ip.ElasticIPState)}
```

Comment:
Random thought: we could get rid of the whole `create_after` ordering by constructing the *actual* dependency graph of resources (rather than an ordering between resource types). I.e., if a resource X has an attribute that refers to `res-Y`, then it has a dependency on Y.

This would eliminate more boring code (`create_after`) and improve parallelism since more resources could be created in parallel.

Reply:
Yeah that would be ideal. Still thinking what's the best way to do that but having a global DAG will make things much better.